spec/s3/app_spec.rb

```diff
@@ -0,0 +1,64 @@
+      with(body: "si").
+      to_return(status: 200, headers: { etag: '"0815etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/aguacate").
+      with(body: "aye").
+      to_return(status: 200, headers: { etag: '"0915etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/aguacate").
+      with(body: "deliciosa").
+      to_return(status: 200, headers: { etag: '"0815etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/aguacate").
+      with(body: "muy deliciosa").
+      to_return(status: 200, headers: { etag: '"0815etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/banano").
+      with(body: "si").
+      to_return(status: 200, headers: { etag: '"0815etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/banano").
+      with(body: "oh, no").
+      to_return(status: 200, headers: { etag: '"0817etag"' })
+    stub_request(:head, "#{container_url_for("phil")}/food/banano").
+      to_return(status: 200, headers: { last_modified: "Fri, 04 Mar 2016 12:20:20 GMT" })
+    stub_request(:head, "#{container_url_for("phil")}/food/aguacate").
+      to_return(status: 200, headers: { last_modified: "Fri, 04 Mar 2016 12:20:18 GMT" })
+    stub_request(:delete, "#{container_url_for("phil")}/food/aguacate").
+      to_return(status: 200, headers: { etag: '"0815etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/camaron").
+      to_return(status: 200, headers: { etag: '"0816etag"' })
+    stub_request(:head, "#{container_url_for("phil")}/food/camaron").
+      to_return(status: 200, headers: { last_modified: "Fri, 04 Mar 2016 12:20:18 GMT" })
+    stub_request(:delete, "#{container_url_for("phil")}/food/camaron").
+      to_return(status: 200, headers: { etag: '"0816etag"' })
+    stub_request(:put, "#{container_url_for("phil")}/food/desayunos/bolon").
+      to_return(status: 200, headers: { etag: '"0817etag"' })
+    stub_request(:head, "#{container_url_for("phil")}/food/desayunos/bolon").
+      to_return(status: 200, headers: { last_modified: "Fri, 04 Mar 2016 12:20:18 GMT" })
+    stub_request(:delete, "#{container_url_for("phil")}/food/desayunos/bolon").
+      to_return(status: 200, headers: { etag: '"0817etag"' })
+    stub_request(:get, "#{container_url_for("phil")}/food/aguacate").
+      to_return(status: 200, body: "rootbody", headers: { etag: '"0817etag"', content_type: "text/plain; charset=utf-8" })
+    stub_request(:put, "#{container_url_for("phil")}/bamboo.txt").
+      to_return(status: 200, headers: { etag: '"0818etag"' })
+    stub_request(:head, "#{container_url_for("phil")}/bamboo.txt").
+      to_return(status: 200, headers: { last_modified: "Fri, 04 Mar 2016 12:20:18 GMT" })
+    stub_request(:head, "#{container_url_for("phil")}/food/steak").
+      to_return(status: 404)
+    stub_request(:get, "#{container_url_for("phil")}/food/steak").
+      to_return(status: 404)
```

Comment:
These stubs are messy af. How is anyone supposed to know what's being stubbed there? Also, there are quite some duplicates with the exact same return values. I think we need to clean this up before merging.

Why is it necessary to break them away from their context entirely in the first place?

Reply:
Them being all defined at once is caused by the way the specs are shared between the two providers, I'm open to suggestions to improve it

Reply:
Yes, at least it should be clear what each spec is for. Otherwise it's all implicit and super hard to both read and change.

Reply:
Yes, at least it should be clear what each spec is for. Otherwise it's all implicit and super hard to both read and change.

Reply:
> Them being all defined at once is caused by the way the specs are shared between the two providers, I'm open to suggestions to improve it

I don't fully understand the reason there. So it's less readable in order to avoid duplication?

Reply:
In a way yes, we're setting up all the request stubs at once for all specs. A lot of the requests are linked (for example PUTs and HEADs on S3 to the same URLs). I'm going to start by adding comments and seeing how I can organize them

Reply:
I have managed to remove some of the stubs, and grouped them together as well as add comment. This is a step in the right direction but it's still not easy to read